src/renderer/atlas/BackendD3D.cpp

```diff
@@ -1104,12 +1108,27 @@ void BackendD3D::_drawTextOverlapSplit(const RenderingPayload& p, u16 y)
     originalLeft = std::max(originalLeft, 0);
     originalRight = std::min(originalRight, cellCount * columnAdvanceInPx);
 
-    auto column = originalLeft / columnAdvanceInPx + 1;
+    if (originalLeft >= originalRight)
+    {
+        return;
+    }
+
+    const auto colors = &p.foregroundBitmap[p.colorBitmapRowStride * y];
+
+    // As explained in the beginning, column and clipLeft should by in multiples of columnAdvance
+    // and columnAdvanceInPx respectively, because that's how line renditions work.
+    auto column = originalLeft / columnAdvanceInPx;
     auto clipLeft = column * columnAdvanceInPx;
     column *= columnAdvance;
 
-    const auto colors = &p.foregroundBitmap[p.colorBitmapRowStride * y];
-    auto lastFg = originalQuad.color;
+    // Our loop below will split the ligature by processing it from left to right.
+    // Some fonts however implement ligatures by replacing a string like "&&" with a whitespace padding glyph,
+    // followed by the actual "&&" glyph which has a 1 column advance width. In that case the originalQuad
+    // will have the .color of the 2nd column and not of the 1st one. We need to fix that here.

```

Comment:
The comment mentions 2-column glyphs.  Has the code been tested with 3-column and 4-column wide glyphs like `<--` and `<!--`?

Reply:
The code doesn't seems to be specific to any column sized glyphs, instead it fixes the scenario where same cell is shared by two adjacent glyphs.

Reply:
In this repository I added a tool called benchcat (bc.exe) a little while ago. Among others it's able to print a text file with a randomized color per extended grapheme. This is the file I use for testing:
```
__      --      ---     -~      ;;      ::      :::     ::=
://     :=      !:      !!      !!.     !.      !=      !==
?:      ??      ??=     ?.      ?=      .-      .?      ..
...     ..<     ..=     .=      (*      [|      ]#      {|
}#      *)      **      ***     */      *>      /*      //
///     /\      />      \/      \\      &&      #_      #_(
#:      #!      #?      #(      #[      #{      #=      %%
^=      ++      +++     +>      <!--    <*      <*>     </
</>     <+      <+>     <<      <<<     <<=     <=      <>
<|      <|>     <||     <|||    <~      <~>     <~~     <$
<$>     =<<     ==      ===     =>>     >=      >>      >>=
>>>     |]      |}      |>      ||      ||>     |||     |||>
~-      ~@      ~=      ~>      ~~      ~~>     $>      www
```

which then looks like this:
![image](https://github.com/microsoft/terminal/assets/2256941/76e6fe91-6253-4c8f-b42a-bf73cc0513fb)

Reply:
Cool and useful tool!  It might be worth having a fixed-color (non-random) mode as well, so that it's possible for a human to verify not only that colors are different, but also that they're the _expected_ color.  Having 3-4 preselected colors with high contrast/hue differences from each other could enhance the ability to quickly detect issues at a glance.

Also, for example the `:=` and `::=` and `!!` and `\/` and `<>` and `<~~` and `=<<` cases got adjacent random colors that are very similar to each other.  It might be useful to skew the random colors to avoid similarities to adjacent colors.  (I have no colorblindness, and I'm able to see differences in all of the cases I listed -- but (a) not everyone will be able to, and (b) the RGB values are pretty close to each other and are thus inherently harder to visually detect quickly.)

Reply:
> The code doesn't seem to be specific to any column sized glyphs, instead it fixes the scenario where same cell is shared by two adjacent glyphs.

@tusharsnx I was asking about the testing validation, not about the code.  The comment in the code explicitly stated "1st" and "2nd" column; there was no mention of 3- or 4-character ligatures.  It was ambiguous what the scope of testing had been, so I asked whether 3- and 4-character ligatures had been tested.  It's an important question to ask, under the circumstances.  Leonard's response is exactly what I was hoping for, and covers the broad spectrum of cases.  ❤️